meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClient.java

```diff
@@ -115,6 +117,41 @@ public void create(String key, Object data, MetaClientInterface.EntryMode mode)
     }
   }
 
+  @Override
+  public void recursiveCreate(String key, T data, EntryMode mode) {
+    recursiveCreateHelper(key, data, mode, -1);
+  }
+
+  @Override
+  public void recursiveCreateWithTTL(String key, T data, long ttl) {
+    recursiveCreateHelper(key, data, EntryMode.TTL, ttl);
+  }
+
+  private void recursiveCreateHelper(String key, T data, EntryMode mode, long ttl) {
+    boolean retry;
+    EntryMode parentMode = (EntryMode.EPHEMERAL.equals(mode) ?
+        EntryMode.PERSISTENT : mode);
+    do {
+      retry = false;
+      try {
+        if (EntryMode.TTL.equals(mode)) {
+          createWithTTL(key, data, ttl);
+        } else {
+          create(key, data, mode);
+        }
+      } catch (MetaClientNoNodeException e) {
+        retry = true;
+        String parentPath = getZkParentPath(key);
+        try {
+          recursiveCreateHelper(parentPath, null, parentMode, ttl);
+        } catch (MetaClientNodeExistsException e1) {
+          return;
```

Comment:
Should we not let this fail silently and return MetaClientException, so that client is aware of failure?

Reply:
Good point. Also after thinking about it more I'm not sure we'd actually ever reach this catch as we only try again when a parent node does not exist so we shouldn't get that error then subsequently get a nodeExistsException. If it does occur somehow it should allowed to bubble up